echo.go

```diff
@@ -606,16 +606,29 @@ func (e *Echo) Start(address string) error {
 
 // StartTLS starts an HTTPS server.
 func (e *Echo) StartTLS(address string, certFile, keyFile string) (err error) {
-	if certFile == "" || keyFile == "" {
+	return e.StartTLSFromFile(address, certFile, keyFile, true)
+}
+
+// StartTLSFromFile starts an HTTPS server.
+// If `fromFile` is `true`: `cert` and `key` are paths to files that contain the certificate and key.
+// If `fromFile` is `false`: `cert` and `key` contain the certificate and key.
+func (e *Echo) StartTLSFromFile(address string, cert, key string, fromFile bool) (err error) {
```

Comment:
The name is misleading. It loads either from file or from the strings provided.

It would be nice if we could avoid adding another method to start TLS... I generally avoid using `interface{}` as much as possible, but can't help wondering if it wouldn't be appropriate here. If we change the signature of `StartTLS()` to `StartTLS(address string, certFile, keyFile interface{})` then the callers could submit either `string` (which would mean path) or `[]byte` (which would mean key/cert content). It would be backward compatible, it would continue to accept strings as input. But it would also support the new feature. Thoughts @mbana @im-kulikov @vishr ?

Either way, this sounds like a very useful feature :-) Now one can feed it straight from AWS Parameter Store or from Vault :-)

Reply:
```go
func (e *Echo) StartTLSFromFile(address string, key string, reader io.Reader) (err error) {
```

mb this is more useful?

Reply:
Hardly @im-kulikov as https://golang.org/pkg/crypto/tls/#Certificate only takes in either path to files or content of certs (but no io.Reader). So asking for that param means both more work for callers and more fork for the function itself, makes little sense.

Reply:
You can read from io.Reader key and cert and pass it to `X509KeyPair`.. 

https://golang.org/pkg/crypto/tls/#X509KeyPair

`func X509KeyPair(certPEMBlock, keyPEMBlock []byte) (Certificate, error) { `

Reply:
something like this:

```go
func (e *Echo) StartTLSFromFile(address string, key, cert io.Reader) (err error) {
```

Reply:
I don't like the idea of using flag to determine what arguments contain

Reply:
`io.Reader` can be `strings.NewReader(<string>)` / `bytes.NewBuffer(<bytes>)` / file / etc.. 

such a decision seems more logical to me

Reply:
@im-kulikov The "we can read from" part is what I meant by more work ;-) More work on the caller to create the `io.Reader` and then more work on the function to read from it, when all is needed is either a `string` or a `[]byte`. Besides, this does not solve the problem with needing an extra param to signal if this is a file or a string (that is going to be transformed into an io.Reader).

Here's what I had in mind:

```Go
func (e *Echo) StartTLS(address string, certFile, keyFile interface{}) (err error) {
    var cert []byte
    switch v := certFile.(type) {
    case string:
        if cert, err = ioutil.ReadFile(certFile); err != nil {
            return
        }
    case []byte:
        cert = certFile
    default:
        return fmt.Errorf("invalid cert type. Can only accept string or []byte")
    }

    // repeat for keyFile

    s := e.TLSServer
	  s.TLSConfig = new(tls.Config)
	  s.TLSConfig.Certificates = make([]tls.Certificate, 1)
		if s.TLSConfig.Certificates[0], err = tls.X509KeyPair(cert, key); err != nil {
        return
    }
	
	return e.startTLS(address)
}
```

No new function, no extra param, backward compatible. If you could show me what you have in mind with `io.Reader` then we can compare and pick the simpler one.

Reply:
thank you. now I see what you mean.

Reply:
Great then :) 

Also, giving it a 2nd look, I would actually introduce an unexported function to clean it up further:

```Go
func (e *Echo) StartTLS(address string, certFile, keyFile interface{}) (err error) {
    var cert []byte
    if cert, err = filepathOrContent(certFile); err != nil {
        return
    }

    var key []byte
    if key, err = filepathOrContent(keyFile); err != nil {
        return
    }

    s := e.TLSServer
    s.TLSConfig = new(tls.Config)
    s.TLSConfig.Certificates = make([]tls.Certificate, 1)
    if s.TLSConfig.Certificates[0], err = tls.X509KeyPair(cert, key); err != nil {
        return
    }
	
    return e.startTLS(address)
}

func filepathOrContent(fileOrContent interface{}) (content []byte, err error) {
    switch v := certFile.(type) {
    case string:
        return ioutil.ReadFile(certFile)
    case []byte:
        return v, nil
    default:
        return nil, fmt.Errorf("invalid cert type. Can only accept string or []byte")
    }
}
```

Yeah, I like this better.

Reply:
sgtm

Reply:
LGTM

Reply:
Thanks @alexaandru that looks good. I was worried about breaking backwards compatibility hence I introduced the new function.